sktree/tree/_unsup_splitter.pxd

```diff
@@ -1,93 +1,42 @@
+    keep count of the "number of samples (weighted)". All samples have a default weight
+    of '1'.
+    """
 
-    # Internal structures
-    cdef public UnsupervisedCriterion criterion      # Impurity criterion
-    cdef public SIZE_t max_features      # Number of features to test
-    cdef public SIZE_t min_samples_leaf  # Min samples in a leaf
-    cdef public double min_weight_leaf   # Minimum weight in a leaf
+    # XXX: not sure if this will work since it subclasses
+    # an existing extension type attribute
+    # cdef public UnsupervisedCriterion criterion
 
-    cdef object random_state             # Random state
-    cdef UINT32_t rand_r_state           # sklearn_rand_r random number state
-
-    cdef SIZE_t[::1] samples             # Sample indices in X, y
-    cdef SIZE_t n_samples                # X.shape[0]
-    cdef double weighted_n_samples       # Weighted number of samples
-    cdef SIZE_t[::1] features            # Feature indices in X
-    cdef SIZE_t[::1] constant_features   # Constant features indices
-    cdef SIZE_t n_features               # X.shape[1]
-    cdef DTYPE_t[::1] feature_values     # temp. array holding feature values
-
-    cdef SIZE_t start                    # Start position for the current node
-    cdef SIZE_t end                      # End position for the current node
-
-    cdef const DOUBLE_t[:] sample_weight
-
-    # The samples vector `samples` is maintained by the Splitter object such
-    # that the samples contained in a node are contiguous. With this setting,
-    # `node_split` reorganizes the node samples `samples[start:end]` in two
-    # subsets `samples[start:pos]` and `samples[pos:end]`.
-
-    # The 1-d  `features` array of size n_features contains the features
-    # indices and allows fast sampling without replacement of features.
-
-    # The 1-d `constant_features` array of size n_features holds in
-    # `constant_features[:n_constant_features]` the feature ids with
-    # constant values for all the samples that reached a specific node.
-    # The value `n_constant_features` is given by the parent node to its
-    # child nodes.  The content of the range `[n_constant_features:]` is left
-    # undefined, but preallocated for performance reasons
-    # This allows optimization with depth-based tree building.
-
-    # Methods
-    cdef int node_reset(
-        self,
-        SIZE_t start,
-        SIZE_t end,
-        double* weighted_n_node_samples
-    ) nogil except -1
-
-    cdef int node_split(
-        self,
-        double impurity,   # Impurity of the node
-        SplitRecord* split,
-        SIZE_t* n_constant_features
-    ) nogil except -1
-
-    cdef void node_value(self, double* dest) nogil
-
-    cdef double node_impurity(self) nogil
-
-cdef class UnsupervisedSplitter(BaseUnsupervisedSplitter):
-    """Abstract interface for supervised splitter."""
+    # feature matrix
+    cdef const DTYPE_t[:, ::1] X
```

Comment:
2D X, 1D Xf

Reply:
?

Reply:
I was just reminding myself that X is 2D and Xf is 1D :smile: